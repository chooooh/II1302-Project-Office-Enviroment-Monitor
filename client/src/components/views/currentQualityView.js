import React from "react";                      //import react framework
import { 
        Button, 
        Navbar,
        Nav, 
        Row, 
        Col, 
        Form, 
        Spinner,
        Jumbotron,
        Card,
        ListGroup,
        Container
    } from 'react-bootstrap';                   // import bootstrap components
import 'bootstrap/dist/css/bootstrap.min.css'   // import bootstrap css
import {currentDateTime} from '../../utils/utils.js'  //import function which returns current date and time from utils

//Objects defining css styling for the Bootstrap components
const formStyle = {
    padding: '10px'
};
const cardTextStyle = {
    color: 'black',
};
const jumbotronStyle = {
    marginTop: '70px',
    
    color: 'black',
};
const rowStyle = {
    margin: '15px'
};
const fontSize = {
    fontSize: '17px',
    fontWeight: 'bold',
};
/**
 * Functions that helps determine if the argument num
 * is a number or NaN
 * 
 * @param {Object} num, is either a number NaN
 * @returns true if arg is a number, else false
 */
const isNumber = (num) => {
    if(num/2) {
        return true
    }
    return false
}

/**
 * This Component is the View component. It dictates the layout of UI and presents
 * all data it receives from its corresponding presenter, currentQuality.js.
 * This component also reacts to user actions such as button clicks and user input. 
 * 
 * @param {Object} data, Object containing all Air Quality related data,
 * @param {String} data.airQualityDate, the date and time which states when carbon and volatile was measured.
 * @param {String} data.carbon,  the current carbon value measured in ppm
 * @param {String} data.volatile, the current amount of volatile gases in the room measured in ppb.
 * @param {String} data.temperatureDate, the date and time which states when temperature was measured.
 * @param {String} data.temperature, the current temperature in the room measured in celcius.
 * @param {String} data.humidityDate, the date and time of when the humidity was measured.
 * @param {String} data.humidity, the current humidity in the room measured in %.
 * @param {Integer} numberOfPeople,  the number of people present in the room.
 * @param {callback} onSubmit,  submits the user input (number of people in the room).  
 * @returns jsx
 */
export const CurrentQualityView = ({data, fetchTime, numberOfPeople, onSubmit}) => {  
    let temperature     = null;
    let humidity        = null;
    let people          = null;
    let volatile        = null;
    let carbon          = null;

    let lastFetch = null;
    if (data != null) lastFetch = currentDateTime(); 

    if(data != null) {
        temperature     = Number.parseFloat(data.temperature).toFixed(2);
        humidity        = Number.parseFloat(data.humidity).toFixed(2);
        volatile        = Number.parseFloat(data.volatile).toFixed(0);
        carbon          = Number.parseFloat(data.carbon).toFixed(0);
    }

    console.log("HUM", humidity)

    const textInput = React.createRef(); 
    const [peeps, setPeople] = React.useState(null);

    console.log("in the view component")
    return (
        <Container fluid>
            <Navbar bg="dark" variant="dark" fixed = "top">
                <Navbar.Brand href="#home">Home</Navbar.Brand>
                <Nav className="mr-auto">
                <Nav.Link href="#history">History</Nav.Link>
                <Nav.Link href="#currentQuality">Current Quality</Nav.Link>
                </Nav>
            </Navbar>
            <Jumbotron style = {jumbotronStyle}>
                <h1>Office Environment Monitor!</h1>
                    <p>
                    On this page you can find all of the current statistics being monitored.<br/>
                        If the number of people displayed on this page is incorrect you can change it to the proper amount below.<br/>
                    </p>
            </Jumbotron>
            <Container fluid data-testid = "air-quality-content">
            {data == null ? <Spinner data-testid = "Spinner" animation="border" role="status"></Spinner>
            :<Card className="text-center">
            <Card.Header style = {cardTextStyle}>Current Air Quality</Card.Header>
            <Card.Body>
                <Card.Title style = {cardTextStyle}>The current measurements are displayed here</Card.Title>
                <Row style = {rowStyle}>
                <Col>
                <Card border = "success">
                    <Card.Body>
                    <Card.Title style = {cardTextStyle}>Humidity</Card.Title>
                        <Card.Subtitle className="mb-2 text-muted" data-testid = "date-humidity">{"Measured: " + data.humidityDate}</Card.Subtitle>
                        <Card.Text style = {cardTextStyle}>
                            <ListGroup variant="flush">
                                <ListGroup.Item data-testid = "actual-humidity" style = {fontSize}>{isNumber(humidity) ? humidity + "%" : "N/A"} </ListGroup.Item>
                            </ListGroup>
                        </Card.Text>
                    </Card.Body>
                </Card>
                </Col>
                <Col>
                <Card border = "success">
                    <Card.Body>
                        <Card.Title style = {cardTextStyle}>Gases</Card.Title>
                        <Card.Subtitle className="mb-2 text-muted" data-testid = "date-air-quality">{"Measured: " + data.gasesDate }</Card.Subtitle>
                        <Card.Text style = {cardTextStyle}>
                            <ListGroup variant="flush">
                                <ListGroup.Item data-testid = "actual-carbon" style = {fontSize}>{isNumber(carbon) ? "Carbon: " + carbon + " ppm" : "N/A"}  </ListGroup.Item>
                                <ListGroup.Item data-testid = "actual-volatile-gases" style = {fontSize}>{isNumber(volatile) ? "Volatile gases: " + volatile + " ppb" : "N/A"}  </ListGroup.Item>
                            </ListGroup>
                        </Card.Text>
                    </Card.Body>
                </Card>
                </Col>
                <Col>
                <Card border = "success" >
                    <Card.Body>
                    <Card.Title style = {cardTextStyle}>Temperature</Card.Title>
                        <Card.Subtitle className="mb-2 text-muted" data-testid = "date-temperature">{"Measured: " + data.temperatureDate}</Card.Subtitle>
                        <Card.Text style = {cardTextStyle}>
                            <ListGroup variant="flush">
                                <ListGroup.Item data-testid = "actual-temperature" style = {fontSize}>{isNumber(temperature) ? temperature + " C" : "N/A"}</ListGroup.Item>
                            </ListGroup>
                        </Card.Text>
                    </Card.Body>
                </Card>
                </Col>
            </Row>
                <Button data-testid = "history-button" variant="primary">History</Button>
            </Card.Body>
            {fetchTime ? <Card.Footer className="text-muted">{"Last fetch: " + fetchTime}</Card.Footer> : <Spinner data-testid = "Spinner" animation="border" role="status"></Spinner>}
            </Card>}
            </Container>
            <Row style = {rowStyle} className = "justify-content-center">
                <Form style = {formStyle} data-testid = "form-component">
                {data == null? <Spinner data-testid = "Spinner" animation="border" role="status"></Spinner> :<Form.Label data-testid = "people">{"Present people: " +  numberOfPeople }</Form.Label>}
                    <Form.Group controlId="formBasicEmail">
                        <Form.Label>Here you can insert the current number of people in the room</Form.Label>
                        <Form.Control placeholder="Enter number of people" ref = {textInput}  onChange = {() => { setPeople(textInput.current.value); console.log(textInput.current.value) }} type = "text"/>
                        <Form.Text className="text-muted">
                        Ensure that the number is correct.
                        </Form.Text>
                    </Form.Group>
                    <Button data-testid = "submit-button" variant="primary" type="submit" onClick = {(event) => {onSubmit(peeps); event.preventDefault()}}>
                        Submit
                    </Button>
                </Form>
            </Row>  
        </Container>
  );
};