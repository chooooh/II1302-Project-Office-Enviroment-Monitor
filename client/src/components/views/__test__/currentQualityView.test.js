import  React from 'react';
import  ReactDOM from 'react-dom';
import {CurrentQualityView} from '../currentQualityView';

import "@testing-library/jest-dom";
import {render, cleanup} from '@testing-library/react';

import renderer from "react-test-renderer";

import {within} from "@testing-library/dom";
import "@babel/plugin-syntax-jsx";
import "@babel/preset-react";

/**
 * After eacht test perform cleanup. 
 * (Prevents rendering the same components multiple times)
 */
afterEach(cleanup);

/**
 * Attempt to render the current quality view component without crashing
 */
it("render CurrentQualityView without crashing", () => {
    //create standard div
    const div = document.createElement("div");
    //try to render the div, attach our view component to that div
    ReactDOM.render( <CurrentQualityView />, div);
})

/**
 * Takes a snapshot of the entire CurrentQualityView component
 */
it("matches snapshot", () => {
    const currentQualityView = renderer.create(<CurrentQualityView/>).toJSON();
    expect(currentQualityView).toMatchSnapshot();
})

/**
 * This test ensures that the submit button displays the correct text
 */
it("renders submit button text correctly", () => {
    const {getByTestId} = render(<CurrentQualityView />);
    expect(getByTestId("submit-button")).toHaveTextContent("Submit");
})
/**
 * This test ensures that the history button displays the correct text
 */
it("renders history button text correctly", () => {
    //Creating a data object to passed to the view component
    //else the cards wont be rendered and a Spinner will be rendered instead.
    //and so this prop must be passed down or else the test wont find the button.
    const data = {
        airQualityDate:  "2021-04-28",
        humidityDate:    "2022-04-28",
        temperatureDate: "2023-04-28",
        carbon: "10",
        volatile: 12,
        humidity: 30,
        temperature: 24 
    }
    const {getByTestId} = render(<CurrentQualityView data = {data}/>);
    expect(getByTestId("history-button")).toHaveTextContent("History");
})

/**
 * This test ensures that when the data prop is sent down to the view containing all the different dates
 * of each measurment. All dates are correctly rendered on the correct places.
 */
describe("Ensuring that date is rendered correctly on all places", () => {
    //Object to be passed as prop when rendering the entire CurrentQualityViewComponent
    //the different dates are used as a timestamp to show when the different reading were done.
    const data = {
        gasesDate:  "2021-04-28",
        humidityDate:    "2022-04-28",
        temperatureDate: "2023-04-28" 
    }

    it("Renders the air quality date correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("date-air-quality")).toHaveTextContent("2021-04-28");
    })
    it("Renders the humidity date correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("date-humidity")).toHaveTextContent("2022-04-28");
    })
    it("Renders the temperature date correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("date-temperature")).toHaveTextContent("2023-04-28");
    })
})

/**
 * This test is written to ensure that if the data regarding volatile gases, carbon
 * humidity and temperature, are not numbers then the view displays N/A instead
 */
describe("This test ensures that carbon, temperature, humidity and volatile gases receives the correct values and renders correctly", () => {
    const data = {
        carbon: "String",
        volatile: "String",
        humidity: "String",
        temperature: "String"
    }
    it("Renders the given carbon correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-carbon")).toHaveTextContent("N/A");
    })
    it("Renders the given volatile gases correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-volatile-gases")).toHaveTextContent("N/A");
    })
    it("Renders the given temperature correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-temperature")).toHaveTextContent("N/A");
    })
    it("Renders the given humidity correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-humidity")).toHaveTextContent("N/A");
    })
})
/**
 * This test ensures that all measured values, such as carbon, volatile, temperature and humidity 
 * which are fetched from the database are correctly rendered in the view and on the correct places.
 */
 describe("This test ensures that carbon, temperature, humidity and volatile gases renders N/A when they dont contain values", () => {
    const data = {
        carbon: 10,
        volatile: 12,
        humidity: 30,
        temperature: 24
    }
    it("Renders the given carbon correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-carbon")).toHaveTextContent("10");
    })
    it("Renders the given volatile gases correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-volatile-gases")).toHaveTextContent("12");
    })
    it("Renders the given temperature correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-temperature")).toHaveTextContent("24");
    })
    it("Renders the given humidity correctly", () => {
        const {getByTestId} = render(<CurrentQualityView data = {data}/>);
        expect(getByTestId("actual-humidity")).toHaveTextContent("30");
    })
})

/**
 * This test ensures that if the data prop, which is passed down from the presenter to the view, is null
 * then a bootstrap spinner component is rendered instead of the main card component.
 */
test('Spinner is rendered instead of the main Card component', () => {
    const data = null;
    const {getAllByTestId, getByTestId } = render(<CurrentQualityView data = {data}/>);
    const date = getByTestId('air-quality-content')
    const spinnerInDate = within(date).getAllByTestId('Spinner')
    expect(spinnerInDate.length).toBe(1);
});

/**
 * This test ensures that if the people data is null then a spinner is rendered instead of the entire
 * boostrap form component.
 */
test('Spinner is rendered instead of the Form component', () => {
    const data = null;
    const {getAllByTestId, getByTestId } = render(<CurrentQualityView data = {data}/>);
    const form = getByTestId('form-component')
    const spinnerInDate = within(form).getAllByTestId('Spinner')
    expect(spinnerInDate.length).toBe(1);
});






