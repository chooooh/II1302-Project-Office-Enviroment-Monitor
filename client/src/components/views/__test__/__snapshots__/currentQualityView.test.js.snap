// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="container-fluid"
>
  <nav
    className="navbar navbar-expand navbar-dark bg-dark fixed-top"
  >
    <a
      className="navbar-brand"
      href="#home"
    >
      Home
    </a>
    <div
      className="mr-auto navbar-nav"
      onKeyDown={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key="#history"
        href="#history"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        History
      </a>
      <a
        className="nav-link"
        data-rb-event-key="#currentQuality"
        href="#currentQuality"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        Current Quality
      </a>
    </div>
  </nav>
  <div
    className="jumbotron"
    style={
      Object {
        "color": "black",
        "marginTop": "70px",
      }
    }
  >
    <h1>
      Office Environment Monitor!
    </h1>
    <p>
      On this page you can find all of the current statistics being monitored.
      <br />
      If the number of people displayed on this page is incorrect you can change it to the proper amount below.
      <br />
    </p>
  </div>
  <div
    className="container-fluid"
    data-testid="air-quality-content"
  >
    <div
      className="spinner-border"
      data-testid="Spinner"
      role="status"
    />
  </div>
  <div
    className="justify-content-center row"
    style={
      Object {
        "margin": "15px",
      }
    }
  >
    <form
      className=""
      data-testid="form-component"
      style={
        Object {
          "padding": "10px",
        }
      }
    >
      <div
        className="spinner-border"
        data-testid="Spinner"
        role="status"
      />
      <div
        className="form-group"
      >
        <label
          className="form-label"
          htmlFor="formBasicEmail"
        >
          Here you can insert the current number of people in the room
        </label>
        <input
          className="form-control"
          id="formBasicEmail"
          onChange={[Function]}
          placeholder="Enter number of people"
          type="text"
        />
        <small
          className="text-muted form-text"
        >
          Ensure that the number is correct.
        </small>
      </div>
      <button
        className="btn btn-primary"
        data-testid="submit-button"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
